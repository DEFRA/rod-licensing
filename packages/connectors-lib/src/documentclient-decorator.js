import db from 'debug'
import { DynamoDB } from '@aws-sdk/client-dynamodb'
import { DynamoDBDocument, BatchWriteCommand, QueryCommand, ScanCommand } from '@aws-sdk/lib-dynamodb'
const debug = db('connectors:aws')

export const createDocumentClient = (options = {}) => {
  const client = new DynamoDB({
    ...options,
    region: options.region || process.env.AWS_REGION || 'eu-west-2'
  })
  const docClient = DynamoDBDocument.from(client)

  // Support for large query/scan operations which return results in pages
  const wrapPagedDocumentClientOperation = (operationName) => {
    return async (params) => {
      const items = []
      let lastEvaluatedKey = null
      do {
        const Command = operationName === 'query' ? QueryCommand : ScanCommand
        const response = await docClient.send(new Command({
          ...params,
          ...(lastEvaluatedKey && { ExclusiveStartKey: lastEvaluatedKey })
        }))
        lastEvaluatedKey = response.LastEvaluatedKey
        response.Items && items.push(...response.Items)
      } while (lastEvaluatedKey)
      return items
    }
  }
  docClient.queryAllPromise = wrapPagedDocumentClientOperation('query')
  docClient.scanAllPromise = wrapPagedDocumentClientOperation('scan')

  /**
   * Handles batch writes which may return UnprocessedItems.  If UnprocessedItems are returned then they will be retried with exponential backoff
   *
   * @param {DocumentClient.BatchWriteItemInput} params as per DynamoDB.DocumentClient.batchWrite
   * @returns {Promise<void>}
   */
  docClient.batchWriteAllPromise = async (params) => {
    let request = { ...params }
    let hasUnprocessedItems = !!Object.keys(request.RequestItems).length
    let unprocessedItemsDelay = 500
    let maxRetries = 10
    while (hasUnprocessedItems) {
      const result = await docClient.send(new BatchWriteCommand(request))
      hasUnprocessedItems = !!Object.keys(result.UnprocessedItems ?? {}).length
      if (hasUnprocessedItems) {
        request = { ...params, RequestItems: result.UnprocessedItems }
        if (maxRetries-- === 0) {
          throw new Error(
            'Failed to write items to DynamoDB using batch write. UnprocessedItems were returned and maxRetries has been reached.'
          )
        }
        await new Promise(resolve => setTimeout(resolve, unprocessedItemsDelay))
        unprocessedItemsDelay = Math.min(2500, unprocessedItemsDelay * 1.5)
        debug('Replaying DynamoDB batchWrite operation due to UnprocessedItems: %o', request.RequestItems)
      }
    }
  }

  docClient.createUpdateExpression = (payload) =>
    Object.entries(payload).reduce(
      (acc, [k, v], idx) => {
        acc.UpdateExpression += `${idx > 0 ? ',' : ''}#${k} = :${k}`
        acc.ExpressionAttributeNames[`#${k}`] = k
        acc.ExpressionAttributeValues[`:${k}`] = v
        return acc
      },
      { UpdateExpression: 'SET ', ExpressionAttributeNames: {}, ExpressionAttributeValues: {} }
    )

  return docClient
}
