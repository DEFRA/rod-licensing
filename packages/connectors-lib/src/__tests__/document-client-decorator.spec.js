import { DynamoDBDocumentClient, QueryCommand, ScanCommand, BatchWriteCommand } from '@aws-sdk/lib-dynamodb'
import { mockClient } from 'aws-sdk-client-mock'
import { createDocumentClient } from '../documentclient-decorator'

describe('document client decorations', () => {
  const ddbMock = mockClient(DynamoDBDocumentClient)
  const docClient = createDocumentClient()

  beforeEach(() => {
    ddbMock.reset()
    // docClient = createDocumentClient()
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('deals with pagination where DynamoDB returns a LastEvaluatedKey in a query response', async () => {
    const testLastEvaluatedKey = { id: '16324258-85-92746491' }

    ddbMock
      .on(QueryCommand)
      .resolvesOnce({
        Items: [{ id: 'item1' }],
        LastEvaluatedKey: testLastEvaluatedKey
      })
      .resolvesOnce({
        Items: [{ id: 'item2' }]
      })

    const items = await docClient.queryAllPromise({ TableName: 'TEST' })

    expect(ddbMock.calls()).toHaveLength(2)
    const firstCall = ddbMock.call(0).args[0].input
    expect(firstCall.TableName).toEqual('TEST')
    const secondCall = ddbMock.call(1).args[0].input
    expect(secondCall.ExclusiveStartKey).toEqual(testLastEvaluatedKey)
    expect(items).toEqual([{ id: 'item1' }, { id: 'item2' }])
  })

  it('deals with pagination where DynamoDB returns a LastEvaluatedKey in a scan response', async () => {
    const testLastEvaluatedKey = { id: '16324258-85-92746491' }

    ddbMock
      .on(ScanCommand)
      .resolvesOnce({
        Items: [{ id: 'item1' }],
        LastEvaluatedKey: testLastEvaluatedKey
      })
      .resolvesOnce({
        Items: [{ id: 'item2' }]
      })

    const items = await docClient.scanAllPromise({ TableName: 'TEST' })

    expect(ddbMock.calls()).toHaveLength(2)
    const firstCall = ddbMock.call(0).args[0].input
    expect(firstCall.TableName).toEqual('TEST')
    const secondCall = ddbMock.call(1).args[0].input
    expect(secondCall.ExclusiveStartKey).toEqual(testLastEvaluatedKey)
    expect(items).toEqual([{ id: 'item1' }, { id: 'item2' }])
  })

  it('deals with UnprocessedItems when making batchWrite requests to DynamoDB', async () => {
    ddbMock
      .on(BatchWriteCommand)
      .resolvesOnce({
        UnprocessedItems: {
          NameOfTableToUpdate: [
            { PutRequest: { Item: { key: '1', field: 'data1' } } },
            { PutRequest: { Item: { key: '2', field: 'data2' } } }
          ]
        }
      })
      .resolvesOnce({
        UnprocessedItems: {}
      })

    const request = {
      RequestItems: {
        NameOfTableToUpdate: [
          { PutRequest: { Item: { key: '1', field: 'data1' } } },
          { PutRequest: { Item: { key: '2', field: 'data2' } } },
          { PutRequest: { Item: { key: '3', field: 'data3' } } }
        ]
      }
    }

    await docClient.batchWriteAllPromise(request)
    expect(ddbMock.calls()).toHaveLength(2)
    const firstCall = ddbMock.call(0).args[0].input
    expect(firstCall.RequestItems.NameOfTableToUpdate).toHaveLength(3)
    const secondCall = ddbMock.call(1).args[0].input
    expect(secondCall.RequestItems.NameOfTableToUpdate).toHaveLength(2)
  })

  it('deals with UnprocessedItems when making batchWrite requests to DynamoDB up to the given retry limit', async () => {
    const batchWriteResponses = Array(11).fill({
      UnprocessedItems: {
        NameOfTableToUpdate: [
          { PutRequest: { Item: { key: '1', field: 'data1' } } },
          { PutRequest: { Item: { key: '2', field: 'data2' } } }
        ]
      }
    })
    ddbMock.on(BatchWriteCommand).resolves(...batchWriteResponses)

    const request = {
      RequestItems: {
        NameOfTableToUpdate: [
          { PutRequest: { Item: { key: '1', field: 'data1' } } },
          { PutRequest: { Item: { key: '2', field: 'data2' } } },
          { PutRequest: { Item: { key: '3', field: 'data3' } } }
        ]
      }
    }
    // Don't delay on setTimeouts!
    jest.spyOn(global, 'setTimeout').mockImplementation(cb => cb())
    await expect(docClient.batchWriteAllPromise(request)).rejects.toThrow(
      'Failed to write items to DynamoDB using batch write. UnprocessedItems were returned and maxRetries has been reached.'
    )
    expect(ddbMock.calls()).toHaveLength(11)
  })

  it('provides a convenience method to simplify building an update expression for DynamoDB', async () => {
    const test = {
      name: 'name-value',
      number: 123
    }
    expect(docClient.createUpdateExpression(test)).toStrictEqual({
      UpdateExpression: 'SET #name = :name,#number = :number',
      ExpressionAttributeNames: {
        '#name': 'name',
        '#number': 'number'
      },
      ExpressionAttributeValues: {
        ':name': 'name-value',
        ':number': 123
      }
    })
  })
})
