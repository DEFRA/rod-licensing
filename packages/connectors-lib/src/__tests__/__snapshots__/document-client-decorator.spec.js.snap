// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document client decorations createUpdateExpression transforms object to an update expression, with provided attribute names and values 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#payload": "payload",
    "#payment": "payment",
    "#permissions": "permissions",
    "#status": "status",
  },
  "ExpressionAttributeValues": Object {
    ":payload": Object {
      "id": "abc-123",
      "name": "ABCDE-123JJ-ABK12",
      "type": "ddd-111-ggg-888",
    },
    ":payment": Object {
      "amount": 16.32,
      "method": "barter",
      "source": "credit",
      "timestamp": "2025-04-09T11:53:17.854Z",
    },
    ":permissions": Array [
      Object {
        "id": "abc-123",
        "name": "ABCDE-123JJ-ABK12",
        "type": "ddd-111-ggg-888",
      },
    ],
    ":status": Object {
      "id": "finalised",
    },
  },
  "UpdateExpression": "SET #payload = :payload,#permissions = :permissions,#status = :status,#payment = :payment",
}
`;
